### ISSUE TITLE
Variable substitution issues in workspace-specific cmake-kits.json

### ISSUE THREAD

**Issue by hellige** [microsoft/vscode-cmake-tools#2737]
### Brief Issue Summary

When using a project-specific kits file in `.vscode/cmake-kits.json`, variable substitution no longer seems to work normally. Given a kit definition like this:

```
  {
    "name": "GCC 11.2.0",
    "compilers": {
      "C": "${workspaceFolder}/../cpp-build-tools/gcc11.2.0-1",
      "CXX": "${workspaceFolder}/../cpp-build-tools/g++11.2.0-1"
    }
  },
```

when opening the project I now see the following lines in the output window:

```
[proc] Executing command: ${workspaceFolder}/../cpp-build-tools/gcc11.2.0-1 -v
[proc] The command: ${workspaceFolder}/../cpp-build-tools/gcc11.2.0-1 -v failed with error: Error: spawn ${workspaceFolder}/../cpp-build-tools/gcc11.2.0-1 ENOENT
...
[proc] The command: gcc11.2 -v failed with error: Error: spawn gcc11.2 ENOENT
[proc] The command: g++11.2 -v failed with error: Error: spawn g++11.2 ENOENT
```

Oddly, the actual cmake configure step still seems to work ok, and find the compiler at the correct path.

In addition (and this was actually the first thing I noticed), I get two new empty directories created _in my home dir_:

```
[~]% ls -ltr ~
...
drwxr-xr-x  2 mhellige mhellige            6 Sep  8 14:16  cpp-build-tools
drwxr-xr-x  2 mhellige mhellige            6 Sep  8 14:16 '${workspaceFolder}'
[~]%
```

I believe the `workspaceFolder` variable should be expanded in `cmake-kits.json`, and in any case the extension should **definitely not** be creating random directories in my home dir!

### CMake Tools Diagnostics

```{
  "os": "linux",
  "vscodeVersion": "1.71.0",
  "cmtVersion": "1.12.26",
  "configurations": [
    {
      "folder": "/home/mhellige/drw/test",
      "cmakeVersion": "3.12.1",
      "configured": true,
      "generator": "Ninja",
      "usesPresets": false,
      "compilers": {
        "C": "/home/mhellige/drw/test/../cpp-build-tools/gcc11.2.0-1",
        "CXX": "/home/mhellige/drw/test/../cpp-build-tools/g++11.2.0-1"
      }
    }
  ],
  "cpptoolsIntegration": {
    "isReady": true,
    "hasCodeModel": true,
    "activeBuildType": "Debug",
    "buildTypesSeen": [
      "Debug"
    ],
    "requests": [],
    "responses": [],
    "partialMatches": [],
    "targetCount": 4,
    "executablesCount": 2,
    "librariesCount": 0,
    "targets": [
      {
        "name": "test_project",
        "type": "UTILITY"
      },
      {
        "name": "fake-target-for-clion",
        "type": "EXECUTABLE"
      },
      {
        "name": "test_project-test",
        "type": "UTILITY"
      },
      {
        "name": "test_project-benchmark",
        "type": "EXECUTABLE"
      }
    ]
  },
  "settings": [
    {
      "communicationMode": "automatic",
      "useCMakePresets": "auto",
      "configureOnOpen": true
    }
  ]
}
```

### Debug Log

```
[main] Safe constructing new CMakeTools instance
[variant] Constructing VariantManager
[main] Constructing new CMakeTools instance
[main] Starting CMakeTools second-phase init
[variant] Loaded new set of variants
[main] CMakeTools instance initialization complete.
[kit] Reading kits file /home/mhellige/.local/share/CMakeTools/cmake-tools-kits.json
[kit] Successfully loaded 10 kits from /home/mhellige/.local/share/CMakeTools/cmake-tools-kits.json
[kit] Reading kits file /home/mhellige/drw/test/.vscode/cmake-kits.json
[kit] Successfully loaded 4 kits from /home/mhellige/drw/test/.vscode/cmake-kits.json
[main] Injecting new Kit into CMake driver
[extension] Configuring workspace on open file:///home/mhellige/drw/test
[main] Starting new CMake driver
[main] Starting CMake driver
[driver] Initializating base driver using kit
[kit] Testing GCC binary: ${workspaceFolder}/../cpp-build-tools/gcc11.2.0-1
[proc] Executing command: ${workspaceFolder}/../cpp-build-tools/gcc11.2.0-1 -v
[proc] The command: ${workspaceFolder}/../cpp-build-tools/gcc11.2.0-1 -v failed with error: Error: spawn ${workspaceFolder}/../cpp-build-tools/gcc11.2.0-1 ENOENT
[kit] Bad GCC binary ("-v" returns -2): ${workspaceFolder}/../cpp-build-tools/gcc11.2.0-1
[driver] CMakeDriver Kit set to GCC 11.2.0
[kit] The environment for kit 'GCC 11.2.0': {
  "ELECTRON_RUN_AS_NODE": "1",
  "USER": "mhellige",
  "LANGUAGE": "en_US",
  "XDG_SEAT": "seat0",
  "TEXTDOMAIN": "im-config",
  "XDG_SESSION_TYPE": "x11",
  "SHLVL": "2",
  "QT4_IM_MODULE": "xim",
  "OLDPWD": "/home/mhellige",
  "HOME": "/home/mhellige",
  "CONDA_SHLVL": "0",
  "DESKTOP_SESSION": "xsession",
  "GTK_MODULES": "gail:atk-bridge",
  "VISUAL": "/usr/bin/vim",
  "COLORTERM": "truecolor",
  "_CE_M": "",
  "IM_CONFIG_PHASE": "2",
  "LOGNAME": "mhellige",
  "GTK_IM_MODULE": "ibus",
  "WINDOWID": "48234499",
  "_": "/usr/bin/code",
  "DEFAULTS_PATH": "/usr/share/gconf/xsession.default.path",
  "XDG_SESSION_ID": "4",
  "TERM": "xterm-256color",
  "USERNAME": "mhellige",
  "_CE_CONDA": "",
  "WINDOWPATH": "2",
  "LC_COLLATE": "POSIX",
  "S_COLORS": "auto",
  "DISPLAY": ":1",
  "HISTSIZE": "20000",
  "LANG": "en_US.utf8",
  "XDG_CURRENT_DESKTOP": "GNOME-Flashback:GNOME",
  "XMODIFIERS": "@im=ibus",
  "XDG_SESSION_DESKTOP": "xsession",
  "DEBIAN_FRONTEND": "noninteractive",
  "SHELL": "/bin/zsh",
  "GDMSESSION": "xsession",
  "QT_ACCESSIBILITY": "1",
  "_JAVA_AWT_WM_NONREPARENTING": "1",
  "TEXTDOMAINDIR": "/usr/share/locale/",
  "QT_IM_MODULE": "ibus",
  "XDG_VTNR": "2",
  "SAVEHIST": "10000",
  "JAVA_HOME": "/site/apps/jdk17",
  "PWD": "/home/mhellige",
  "CLUTTER_IM_MODULE": "xim",
  "XDG_CONFIG_DIRS": "/etc/xdg/xdg-xsession:/etc/xdg",
  "XDG_DATA_DIRS": "/usr/share/xsession:/usr/local/share/:/usr/share/:/var/lib/snapd/desktop",
  "AUTOJUMP_SOURCED": "1",
  "HISTFILE": "/home/mhellige/.zsh_history",
  "VTE_VERSION": "5202",
  "JDK_HOME": "/site/apps/jdk17",
  "EDITOR": "/usr/bin/vim",
  "VSCODE_CWD": "/home/mhellige",
  "APPLICATION_INSIGHTS_NO_DIAGNOSTIC_CHANNEL": "1",
  "VSCODE_CLI": "1",
  "ELECTRON_NO_ATTACH_CONSOLE": "1",
  "CHROME_DESKTOP": "code-url-handler.desktop",
  "ORIGINAL_XDG_CURRENT_DESKTOP": "GNOME-Flashback:GNOME",
  "GDK_BACKEND": "x11",
  "VSCODE_NLS_CONFIG": "{\"locale\":\"en-us\",\"availableLanguages\":{},\"_languagePackSupport\":true}",
  "VSCODE_CODE_CACHE_PATH": "/home/mhellige/.config/Code/CachedData/784b0177c56c607789f9638da7b6bf3230d47a8c",
  "VSCODE_IPC_HOOK": "/run/user/7043/vscode-b24e4f06-1.71.0-main.sock",
  "VSCODE_PID": "17745",
  "VSCODE_AMD_ENTRYPOINT": "vs/workbench/api/node/extensionHostProcess",
  "VSCODE_HANDLES_UNCAUGHT_ERRORS": "true"
}
[driver] Trying to detect generator supported by system
[cms-driver] Run doRefreshExpansions
[cms-client] Started new CMake Server instance with PID 7314
[cache] Reading CMake cache file /home/mhellige/drw/test/build/CMakeCache.txt
[cache] Parsing CMake cache string
[driver] Setting new variant Debug
[cms-driver] Run doRefreshExpansions
[cms-client] Started new CMake Server instance with PID 7331
[cache] Reading CMake cache file /home/mhellige/drw/test/build/CMakeCache.txt
[cache] Parsing CMake cache string
[main] Configuring folder: test 
[main] Saving open files before configure/build
[cms-driver] Run doRefreshExpansions
[driver] Start configure 
[driver] Running pre-configure checks and steps
[driver] Using compilers in GCC 11.2.0 for configure
[cms-driver] Run doRefreshExpansions
[cmake] Using multiarch at x86_64-linux-gnu
[cmake] The test_project compiler supports the C++ 2a standard
[cmake] Configuring done
[cmake] Generating done
[proc] The command: gcc11.2 -v failed with error: Error: spawn gcc11.2 ENOENT
[proc] The command: g++11.2 -v failed with error: Error: spawn g++11.2 ENOENT
[cache] Reading CMake cache file /home/mhellige/drw/test/build/CMakeCache.txt
[cache] Parsing CMake cache string
[cache] Reading CMake cache file /home/mhellige/drw/test/build/CMakeCache.txt
[cache] Parsing CMake cache string
[cache] Reading CMake cache file /home/mhellige/drw/test/build/CMakeCache.txt
[cache] Parsing CMake cache string
```

**Comment by bobbrow** [comment]
The fix should be available in tomorrow's pre-release: 1.13.7

**Comment by hellige** [comment]
I've verified that this is fixed in 1.13.7. Thanks!!

One strange behavior remains. With 1.13.7, my output from above becomes:
```
[proc] Executing command: /home/mhellige/drw/test/../cpp-build-tools/gcc11.2.0-1 -v
... [ everything is fine, cmake runs ok, etc. then finally: ]
[proc] The command: gcc11.2 -v failed with error: Error: spawn gcc11.2 ENOENT
[proc] The command: g++11.2 -v failed with error: Error: spawn g++11.2 ENOENT
```

(And the stray directories are no longer created in my home dir.)

I still can't figure out those last two `ENOENT` lines... Where in the world is `gcc11.2` coming from? My compiler binary is called `gcc11.2.0-1`, as you can see above. I can't find that string in any file in the workspace, nor in `~/.vscode` or `~/.local/share/CMakeTools`. I've tried enabling trace logging and everything else I can think of, but I can't for the life of me figure out where it's getting the string `gcc11.2`...

I'm not sure if this is a different bug, or a configuration issue, or what, but I would be very grateful for any insight!
